lib/posix-procfs/procfs_vnops.c

```diff
@@ -0,0 +1,632 @@
+	rc = uk_store_get_value(entry, u64, &tot_nb_allocs);
+
+	entry = uk_store_get_entry(libukalloc, NULL, "tot_nb_frees");
+	rc = uk_store_get_value(entry, u64, &tot_nb_frees);
+
+	entry = uk_store_get_entry(libukalloc, NULL, "cur_nb_allocs");
+	rc = uk_store_get_value(entry, s64, &cur_nb_allocs);
+
+	entry = uk_store_get_entry(libukalloc, NULL, "max_nb_allocs");
+	rc = uk_store_get_value(entry, s64, &max_nb_allocs);
+
+	entry = uk_store_get_entry(libukalloc, NULL, "cur_mem_use");
+	rc = uk_store_get_value(entry, s64, &cur_mem_use);
+
+	entry = uk_store_get_entry(libukalloc, NULL, "max_mem_use");
+	rc = uk_store_get_value(entry, s64, &max_mem_use);
+
+	entry = uk_store_get_entry(libukalloc, NULL, "nb_enomem");
+	rc = uk_store_get_value(entry, u64, &nb_enomem);
+
+
+	sprintf(to_read, 
+            "last_alloc_size:\t%ld\n"
+            "max_alloc_size:\t\t%ld\n"
+            "min_alloc_size:\t\t%ld\n"
+            "tot_nb_allocs:\t\t%ld\n"
+            "tot_nb_frees:\t\t%ld\n"
+            "cur_nb_allocs:\t\t%ld\n"
+            "max_nb_allocs:\t\t%ld\n"
+            "cur_mem_use:\t\t%ld\n"
+            "max_mem_use:\t\t%ld\n"
+            "nb_enomem:\t\t%ld\n",
+			last_alloc_size,
+			max_alloc_size,
+			min_alloc_size,
+			tot_nb_allocs,
+			tot_nb_frees,
+			cur_nb_allocs,
+			max_nb_allocs,
+			cur_mem_use,
+			max_mem_use,
+			nb_enomem
+			);
+	return to_read;
+}
+
+#endif /* CONFIG_LIBPROCFS_MEMINFO */
+
+#ifdef CONFIG_LIBPROCFS_VERSION
+static int read_version(char *buff, int max_size)
+{
+	const struct uk_store_entry *entry;
+	int rc;
+
+	entry = uk_store_get_entry(libukboot, NULL, "uk_version");
+	rc = uk_store_get_ncharp(entry, buff, max_size);
+
+	return rc;
+}
+#endif /* CONFIG_LIBPROCFS_VERSION */
+
+
+
+
+#ifdef CONFIG_LIBPROCFS_CMDLINE
+/* it requires the structure and lib-newlibc */
+static int read_cmdline(char *buff, int max_size)
+{
+	const struct uk_store_entry *entry;
+	int rc;
+
+	entry = uk_store_get_entry(libukboot, NULL, "uk_arguments");
+	rc = uk_store_get_ncharp(entry, buff, max_size);
+
+	return rc;
+}
+#endif /* CONFIG_LIBPROCFS_CMDLINE */
+
+
+
+#ifdef CONFIG_LIBPROCFS_FIELSYSTEMS
+static int read_filesystems(char *buff, int max_size)
+{
+	const struct uk_store_entry *entry;
+	int rc;
+
+	entry = uk_store_get_entry(libvfscore, NULL, "uk_filesystems");
+	rc = uk_store_get_ncharp(entry, buff, max_size);
+
+	return rc;
+}
+#endif /* CONFIG_LIBPROCFS_FIELSYSTEMS */
+
+#ifdef CONFIG_LIBPROCFS_MOUNTS
+static int read_mounts(char *buff, int max_size)
+{
+	const struct uk_store_entry *entry;
+	int rc;
+
+	entry = uk_store_get_entry(libvfscore, NULL, "uk_mounts");
+	rc = uk_store_get_ncharp(entry, buff, max_size);
+
+	return rc;
+}
+#endif /* CONFIG_LIBPROCFS_MOUNTS */
+
+static int
+procfs_read(struct vnode *vp, struct vfscore_file *fp __unused,
+	   struct uio *uio, int ioflag __unused)
+{
+	struct procfs_node *np =  vp->v_data;
+	size_t len;
+	char *return_me;
+	char *value = malloc(sizeof(char) * 100);
```

Comment:
On error, the buffer will be leaked. Also the size seems a bit random. Probably, it would be better to let the `read_` functions allocate the buffer. They know how large their output will be. You can use a double pointer for `value`.